plugin/src/main/java/org/opensearch/ml/action/prediction/TransportPredictionTaskAction.java

```diff
@@ -180,6 +186,67 @@ public void onFailure(Exception e) {
         }
     }
 
+    private void deployAndPredictRemoteModel(
+        String modelId,
+        ActionListener<MLTaskResponse> wrappedListener,
+        MLPredictionTaskRequest mlPredictionTaskRequest
+    ) {
+        String[] workerNodes = mlModelManager.getWorkerNodes(modelId, FunctionName.REMOTE, true);
+        if (workerNodes != null && workerNodes.length != 0) {
+            // transport layer deployment finished
+            executePredict(mlPredictionTaskRequest, wrappedListener, modelId);
+            return;
+        }
+        mlModelManager.getModel(modelId, ActionListener.wrap(mlModel -> {
+            // Obtain the deployment plan
+            String[] workerNodeIds = mlModel.getPlanningWorkerNodes();
+            MLDeployModelRequest deployModelRequest = MLDeployModelRequest
+                .builder()
+                .modelId(modelId)
+                .async(false)
+                .dispatchTask(false)
+                .modelNodeIds(workerNodeIds)
+                .build();
+            deployAndPredict(modelId, deployModelRequest, mlPredictionTaskRequest, wrappedListener);
+        }, exception -> {
+            log.error("Failed to find the remote model in the model index " + modelId, exception);
+            wrappedListener.onFailure(exception);
+        }));
+    }
+
+    private void deployAndPredict(
+        String modelId,
+        MLDeployModelRequest deployModelRequest,
+        MLPredictionTaskRequest mlPredictionTaskRequest,
+        ActionListener<MLTaskResponse> wrappedListener
+    ) {
+        ActionListener<MLDeployModelResponse> deployModelActionListener = ActionListener.wrap(deployModelResponse -> {
+            // Deployment failed, existing
+            if (!deployModelResponse.getStatus().equals(MLTaskState.COMPLETED.name())) {
+                wrappedListener
+                    .onFailure(
+                        new IllegalArgumentException(
+                            "Deployment Failed for " + modelId + ", please check your remote model before predicting."
+                        )
+                    );
+                return;
+            }
+            String[] workerNodes = mlModelManager.getWorkerNodes(modelId, FunctionName.REMOTE, true);
+            if (workerNodes == null || workerNodes.length == 0) {
+                // The deployment is kicked off, but not finished yet in this node
+                mlModelManager.deployRemoteModelToLocal(modelId, ActionListener.wrap(r -> {
+                    executePredict(mlPredictionTaskRequest, wrappedListener, modelId);
+                }, exception -> {
+                    log.error("Failed to deploy model to local node" + modelId, exception);
+                    wrappedListener.onFailure(exception);
+                }));
+            } else {
+                executePredict(mlPredictionTaskRequest, wrappedListener, modelId);
```

Comment:
So if local `workerNodes` is not null and empty, we will run predict request. If I remember correctly, the predict request could be routed to other nodes, should we make sure the `mlPredictionTaskRequest` only run on local ?

Reply:
This is actually the normal predict case when the model is deployed successfully, so it needs to route the request to the work nodes through RoundRobin. In the case of local deployment, I have changed the mlPredictionTaskRequest to not dispatch tasks so it invokes the prediction only locally.